### PR TITLE
[POCRice - week1] - Completed Week1 Mini Project

### PR DESCRIPTION
**Pull request by kayleabox**
This PR has the exercises for week 1 of the class.

To run merge:
``python week1/run_merge.py``

To run tests for merge function:
``python week1/run_tests.py``

### CODE REVIEW

File: week1/merge.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    return merged_line;
+
+def try_merge(line_wo_zeros, merged_line):
+    """
+    Attempts to merge each number in an array with the one after it.
+    If the number was already merged with the one before it, 
+    no merge will occur.
+    """
+    merged_w_previous = False
+    for index in range(len(line_wo_zeros)):
```
Comment (inchoate) [inline week1/merge.py]: More Pythonic is:
```python
for value, index in enumerate(line_wo_zeros):
  ... do stuff
```

File: week1/merge.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+def merge(line):
+    """
+    Function that merges a single row or column in 2048.
+    """
+    line_without_zeros = [value for value in line if value != 0]
+    merged_line = []
+    
+    try_merge(line_without_zeros, merged_line)
+    while len(line) - len(merged_line) > 0:
```
Comment (inchoate) [inline week1/merge.py]: I think line 10 and this loop can be simplified. How? :-)

File: week1/merge.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        return True
+    else:
+        return False
+
+def append_to_line_if_not_merged(merged_with_previous, merged_line, line_without_zeros, index):
+    """
+    Appends numbers that are not eligible for merging with their
+    neighbors to the placeholder array.
+    """
+    if merged_with_previous != True:
```
Comment (inchoate) [inline week1/merge.py]:
```
if not merged_with_previous:
   ... do stuff
```

Comment (inchoate) [reply week1/merge.py]: You're also not supposed to pass booleans into functions.

Also, you don't need to test whether a boolean is True/False -- just test the value itself! :-)

```python
a = True

if a == True:   # this is an expression that evaluates to `True` which `a` already is.
```

Got it?  :-)

**Review by inchoate — COMMENTED** [review]
A couple small comments. Nice start. It's looking relatively Pythonic, too.

**Comment by inchoate** [comment]
I also recommend installing nose, rednose, and coverage for best practices. Lemme know if you need help on this.

File: week1/merge.py (hunk 3 above)
Comment (kayleabox) [reply week1/merge.py]: eek... I should know this. Uncle Bob taught me better. :/

File: week1/merge.py (hunk 2 above)
Comment (kayleabox) [reply week1/merge.py]: I ended up doing something like this... 
```
[0 for dummy_num in range(len(line) - len(line_without_zeros))]
```
Is that close to what you were thinking?
Also, the class I am taking suggests naming unused variables in python for loops dummy_. Is there a better way to handle this situation?

File: week1/merge.py [hunk 4]
```diff
@@ -7,10 +7,8 @@ def merge(line):
     Function that merges a single row or column in 2048.
     """
     line_without_zeros = [value for value in line if value != 0]
-    merged_line = try_merge(line_without_zeros)
-    while len(line) - len(merged_line) > 0: #still need to fix this guy
-        merged_line.append(0)
-    return merged_line;
+    return (try_merge(line_without_zeros) +
+           [0 for dummy_num in range(len(line) - len(line_without_zeros))])
```
Comment (inchoate) [inline week1/merge.py]: You took my advice. You went a _little_ farther than I would have taken it. For the sake of readability I would have left `merged_line` in there, something like this:
```python
non_zeros = [value for value in line if value != 0]
zeros = [0 for dummy_num in range(len(line) - len(non_zeros))]
return try_merge(non_zeros) + zeros
```

**Review by inchoate — APPROVED** [review]
LGTM

File: week1/merge.py (hunk 4 above)
Comment (kayleabox) [reply week1/merge.py]: :) It is possible I got a little carried away. I am running into something interesting when I try to implement your suggestion. It is not returning an array with 0, instead it is returning an empty array when I try to assign it to the variable zeros. Does this make sense to you? 
```
#this adds 0's to an array
merged_line = try_merge(non_zeros) + [0 for dummy_num in range(len(line) - len(non_zeros))]

#this appears to return an empty array in the program
zeros = [0 for dummy_num in range(len(line) - len(non_zeros))]
```

I might be doing something strange...

Comment (kayleabox) [reply week1/merge.py]: Just kidding! The order of operations was off. This seems to fix it:
```
def merge(line):
    """
    Function that merges a single row or column in 2048.
    """
    non_zeros = try_merge([value for value in line if value != 0])
    zeros = [0 for dummy_num in range(len(line) - len(non_zeros))]
    return (non_zeros + zeros)
```

**Pull request merged**